mm2src/coins/eth.rs

```diff
@@ -2879,30 +2886,42 @@ impl EthCoin {
                     Token::Address(*token_addr),
                     Token::Address(receiver_addr),
                     Token::FixedBytes(secret_hash),
-                    Token::Uint(U256::from(args.time_lock))
+                    Token::Uint(time_lock)
                 ]));
-                let value = U256::from(args.watcher_reward.unwrap_or(0));
+                // wait for the approval tx confirmation only half the time required for the actual htlc tx confirmation
+                let wait_for_approval_confirmation_until = args.wait_for_confirmation_until / 2;
 
                 let arc = self.clone();
                 Box::new(allowance_fut.and_then(move |allowed| -> EthTxFut {
-                    if allowed < value {
+                    if allowed < trade_amount {
                         Box::new(
                             arc.approve(swap_contract_address, U256::max_value())
-                                .and_then(move |_approved| {
-                                    arc.sign_and_send_transaction(
-                                        value,
-                                        Action::Call(swap_contract_address),
-                                        data,
-                                        U256::from(ETH_GAS),
+                                .and_then(move |approved| {
+                                    // make sure the approve tx is confirmed before sending the htlc tx
+                                    arc.wait_for_confirmations(
```

Comment:
I decided to wait for confirmation of the approve transaction, alternatively, we can loop over the allowance fn and check until the amount is allowed. What do you think is the better solution @caglaryucekaya? checking allowance fn might be a cheaper call but I am not sure.

Reply:
https://www.quicknode.com/docs/ethereum/api_credits

Quicknode has an API credit measure here for each method based on its complexity. `wait_for_confirmations`requires two calls at each iteration of the loop: `eth_getTransactionReceipt` (2 credits) and `eth_blockNumber` (1 credit). Allowance query requires one call `eth_call` which is 2 credits. So allowance call is less credits in total according to this list, and one call is better than having two separate calls in general. I think it's worth changing it to a loop over the allowance function.